src/ir/comp.rs

```diff
@@ -1694,6 +1718,71 @@ impl CompInfo {
 
         true
     }
+
+    // Check whether a bit field's bit width expression contains any references
+    // to template parameters. If so, we don't evaluate its width.
+    fn is_field_bit_width_template_dependent(cursor: &clang::Cursor) -> bool {
+        use clang_sys::*;
+
+        if !cursor.is_bit_field() {
+            return false;
+        }
+
+        fn visitor(
+            found_template_parameter: &mut bool,
+            cur: clang::Cursor,
+        ) -> CXChildVisitResult {
+            // The next expression or literal is the bit width. Search it for
+            // references to template parameters.
+            if let Some(referenced) = cur.referenced() {
+                match referenced.kind() {
+                    // If we found a template argument, it is dependent.
+                    CXCursor_TemplateTemplateParameter |
+                    CXCursor_TemplateTypeParameter |
+                    CXCursor_NonTypeTemplateParameter => {
+                        *found_template_parameter = true;
+                        return CXChildVisit_Break;
+                    }
+                    // If we found a type alias or typedef, follow it and check
+                    // there.
+                    CXCursor_TypedefDecl |
+                    CXCursor_TypeAliasDecl |
+                    CXCursor_TypeAliasTemplateDecl => {
+                        referenced.visit(|cur| {
+                            visitor(found_template_parameter, cur)
+                        });
+                        if *found_template_parameter {
+                            return CXChildVisit_Break;
+                        }
+                    }
+                    _ => (),
+                }
+            }
+
+            CXChildVisit_Recurse
+        }
+
+        let mut found_template_parameter = false;
+
+        cursor.visit(|cur| {
+            // The first child may or may not be a TypeRef, depending on whether
+            // the field's type is builtin or not. Skip this.
+            if cur.kind() == CXCursor_TypeRef {
+                return CXChildVisit_Continue;
+            }
+
+            // The next expression or literal is the bit width. Search for
+            // references to template parameters.
+            cur.visit(|next_cur| {
+                visitor(&mut found_template_parameter, next_cur)
+            });
+
+            // There should be no more children.
+            CXChildVisit_Break
+        });
+
+        found_template_parameter
```

Comment:
This seems rather complicated for what's arguably a bug in libclang. In the past I've sent patches there (in the pre-github era) and they were accepted. Would there be any chance to at least send a fix upstream too so that we can remove this?

Reply:
The issue is, to get the fix we’d have to bump the min libclang version to 15. If we don’t then this is still a soundness issue. While this triggers an llvm assert if enabled, it causes a use after return as detected by asan otherwise.

Reply:
Following up again. I do have an upstream change: https://reviews.llvm.org/D130303. However, I think a workaround is still needed here

Essentially, clang_getFieldDeclBitWidth has an undocumented soundness precondition. It is unsound to call it if the cursor points to a bitfield with a template-dependent bit width.

Even when the upstream change lands, this soundness precondition applies to all earlier libclang revisions. Seems like a minefield for users.

I'm open to other ways of implementing this. For example, another option may be to traverse to the bitfield width expression and call clang_Cursor_Evaluate on it. This function already has a check for template dependence: https://github.com/llvm-mirror/clang/blame/aa231e4be75ac4759c236b755c57876f76e3cf05/tools/libclang/CIndex.cpp#L3803

Funnily enough, this check was added years after the function was added. So the same API flaw existed there.